README.md
#  The guardian soccer spider

Gathering soccer summaries from the Guardian

CHANGELOG.md

<a name="0.1.0-8c708a7"></a>
## [0.1.0-8c708a7](https://github.com/compare/v0.1.0-1af124b...0.1.0-8c708a7) (2021-04-12)


<a name="v0.1.0-1af124b"></a>
## [v0.1.0-1af124b](https://github.com/compare/v0.1.0-7bbc4f0...v0.1.0-1af124b) (2021-04-12)


<a name="v0.1.0-7bbc4f0"></a>
## [v0.1.0-7bbc4f0](https://github.com/compare/v0.1.0-0a7af70...v0.1.0-7bbc4f0) (2021-04-12)


<a name="v0.1.0-0a7af70"></a>
## [v0.1.0-0a7af70](https://github.com/compare/v0.1.0-7873116...v0.1.0-0a7af70) (2021-04-12)


<a name="v0.1.0-7873116"></a>
## [v0.1.0-7873116](https://github.com/compare/v0.1.0-365792f...v0.1.0-7873116) (2021-04-11)


<a name="v0.1.0-365792f"></a>
## [v0.1.0-365792f](https://github.com/compare/v0.1.0-3fcde63...v0.1.0-365792f) (2021-04-11)


<a name="v0.1.0-3fcde63"></a>
## [v0.1.0-3fcde63](https://github.com/compare/v0.1.0-9283eac...v0.1.0-3fcde63) (2021-04-11)


<a name="v0.1.0-9283eac"></a>
## [v0.1.0-9283eac](https://github.com/compare/v0.1.0-027382b...v0.1.0-9283eac) (2021-04-11)


<a name="v0.1.0-027382b"></a>
## [v0.1.0-027382b](https://github.com/compare/v0.1.0-136b169...v0.1.0-027382b) (2021-04-11)


<a name="v0.1.0-136b169"></a>
## [v0.1.0-136b169](https://github.com/compare/v0.1.0-550c217...v0.1.0-136b169) (2021-04-11)


<a name="v0.1.0-550c217"></a>
## [v0.1.0-550c217](https://github.com/compare/v0.1.0-42fb947...v0.1.0-550c217) (2021-04-11)


<a name="v0.1.0-42fb947"></a>
## [v0.1.0-42fb947](https://github.com/compare/v0.1.0-addd024...v0.1.0-42fb947) (2021-04-11)


<a name="v0.1.0-addd024"></a>
## [v0.1.0-addd024](https://github.com/compare/v0.1.0-178c7e2...v0.1.0-addd024) (2021-04-11)


<a name="v0.1.0-178c7e2"></a>
## [v0.1.0-178c7e2](https://github.com/compare/v0.1.0-10f02f1...v0.1.0-178c7e2) (2021-04-11)


<a name="v0.1.0-10f02f1"></a>
## [v0.1.0-10f02f1](https://github.com/compare/v0.1.0-f11ec4f...v0.1.0-10f02f1) (2021-04-11)


<a name="v0.1.0-f11ec4f"></a>
## [v0.1.0-f11ec4f](https://github.com/compare/v0.1.0-8856bd4...v0.1.0-f11ec4f) (2021-04-11)


<a name="v0.1.0-8856bd4"></a>
## [v0.1.0-8856bd4](https://github.com/compare/v0.1.0-eba6c9f...v0.1.0-8856bd4) (2021-04-11)


<a name="v0.1.0-eba6c9f"></a>
## [v0.1.0-eba6c9f](https://github.com/compare/v0.1.0-f28dc3d...v0.1.0-eba6c9f) (2021-04-11)


<a name="v0.1.0-f28dc3d"></a>
## [v0.1.0-f28dc3d](https://github.com/compare/v0.1.0-f887773...v0.1.0-f28dc3d) (2021-04-11)


<a name="v0.1.0-f887773"></a>
## [v0.1.0-f887773](https://github.com/compare/v0.1.0-37e4213...v0.1.0-f887773) (2021-04-11)


<a name="v0.1.0-37e4213"></a>
## [v0.1.0-37e4213](https://github.com/compare/v0.1.0-8e5a37d...v0.1.0-37e4213) (2021-04-10)


<a name="v0.1.0-8e5a37d"></a>
## [v0.1.0-8e5a37d](https://github.com/compare/v0.1.0-1f99fc2...v0.1.0-8e5a37d) (2021-04-10)


<a name="v0.1.0-1f99fc2"></a>
## [v0.1.0-1f99fc2](https://github.com/compare/v0.1.0-761c783...v0.1.0-1f99fc2) (2021-04-10)


<a name="v0.1.0-761c783"></a>
## [v0.1.0-761c783](https://github.com/compare/v0.1.0-5f4e3a4...v0.1.0-761c783) (2021-04-10)


<a name="v0.1.0-5f4e3a4"></a>
## [v0.1.0-5f4e3a4](https://github.com/compare/v0.1.0-871adf7...v0.1.0-5f4e3a4) (2021-04-10)


<a name="v0.1.0-871adf7"></a>
## v0.1.0-871adf7 (2021-04-10)


matchreports/matchreports/items.py
# Define here the models for your scraped items
#
# See documentation in:
# https://docs.scrapy.org/en/latest/topics/items.html

import scrapy
from scrapy import Field
from itemloaders.processors import Join, MapCompose, TakeFirst
from datetime import datetime


def ymd(date: datetime):
    return date.strftime("%Y-%m-%d")


class MatchreportsItem(scrapy.Item):
    link = Field(output_processor=TakeFirst())
    headline = Field(output_processor=TakeFirst())
    home_team = Field(output_processor=TakeFirst())
    away_team = Field(output_processor=TakeFirst())
    match_date = Field(input_processor=MapCompose(ymd), output_processor=TakeFirst())
    pictures = Field(output_processor=TakeFirst())
    author = Field(output_processor=TakeFirst())
    stadium = Field(output_processor=TakeFirst())
    text = Field()
    scrape_date = Field(input_processor=MapCompose(ymd), output_processor=TakeFirst())

matchreports/matchreports/settings.py
"""This module contains all the software (not model) specific settings
"""
import os
from typing import List
from pydantic import (
    BaseSettings,
    AnyHttpUrl,
    SecretStr,
    HttpUrl,
)
import pytz


class Settings(BaseSettings):
    LOCAL_TZ = pytz.timezone("Europe/Copenhagen")
    BUCKET_NAME: str


class ScraperSettings(BaseSettings):
    Hello = "No"


class GuadianSiteSettings(BaseSettings):
    BASE_URL = "https://www.theguardian.com/football/premierleague+tone/matchreports"


settings = Settings()
scrabersettings = ScraperSettings()
guardiansitesettings = GuadianSiteSettings()

# Scrapy settings for matchreports project
#
# For simplicity, this file contains only settings considered important or
# commonly used. You can find more settings consulting the documentation:
#
#     https://docs.scrapy.org/en/latest/topics/settings.html
#     https://docs.scrapy.org/en/latest/topics/downloader-middleware.html
#     https://docs.scrapy.org/en/latest/topics/spider-middleware.html

BOT_NAME = "matchreports"

SPIDER_MODULES = ["matchreports.spiders"]
NEWSPIDER_MODULE = "matchreports.spiders"

# Crawl responsibly by identifying yourself (and your website) on the user-agent
# USER_AGENT = 'matchreports (+http://www.yourdomain.com)'

# Obey robots.txt rules
ROBOTSTXT_OBEY = True
USER_AGENTS = [
    ("Mozilla/5.0 (X11; Linux x86_64) " "AppleWebKit/537.36 (KHTML, like Gecko) " "Chrome/57.0.2987.110 " "Safari/537.36"),  # chrome
    ("Mozilla/5.0 (X11; Linux x86_64) " "AppleWebKit/537.36 (KHTML, like Gecko) " "Chrome/61.0.3163.79 " "Safari/537.36"),  # chrome
    ("Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:55.0) " "Gecko/20100101 " "Firefox/55.0"),  # firefox
    ("Mozilla/5.0 (X11; Linux x86_64) " "AppleWebKit/537.36 (KHTML, like Gecko) " "Chrome/61.0.3163.91 " "Safari/537.36"),  # chrome
    ("Mozilla/5.0 (X11; Linux x86_64) " "AppleWebKit/537.36 (KHTML, like Gecko) " "Chrome/62.0.3202.89 " "Safari/537.36"),  # chrome
    ("Mozilla/5.0 (X11; Linux x86_64) " "AppleWebKit/537.36 (KHTML, like Gecko) " "Chrome/63.0.3239.108 " "Safari/537.36"),  # chrome
]
DOWNLOADER_MIDDLEWARES = {
    "scrapy_useragents.downloadermiddlewares.useragents.UserAgentsMiddleware": 500,
}
SPIDER_MIDDLEWARES = {
    "matchreports.middlewares.DeltaFetch": 100,
}
DELTAFETCH_ENABLED = True
COOKIES_ENABLED = False
# Configure maximum concurrent requests performed by Scrapy (default: 16)
CONCURRENT_REQUESTS = 16


# Configure a delay for requests for the same website (default: 0)
# See https://docs.scrapy.org/en/latest/topics/settings.html#download-delay
# See also autothrottle settings and docs
DOWNLOAD_DELAY = 2

# The download delay setting will honor only one of:
# CONCURRENT_REQUESTS_PER_DOMAIN = 16
# CONCURRENT_REQUESTS_PER_IP = 16


# Disable Telnet Console (enabled by default)
# TELNETCONSOLE_ENABLED = False

# Override the default request headers:
# DEFAULT_REQUEST_HEADERS = {
#   'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
#   'Accept-Language': 'en',
# }

# Enable or disable spider middlewares
# See https://docs.scrapy.org/en/latest/topics/spider-middleware.html
# SPIDER_MIDDLEWARES = {
#    'matchreports.middlewares.MatchreportsSpiderMiddleware': 543,
# }

# Enable or disable downloader middlewares
# See https://docs.scrapy.org/en/latest/topics/downloader-middleware.html
# DOWNLOADER_MIDDLEWARES = {
#    'matchreports.middlewares.MatchreportsDownloaderMiddleware': 543,
# }

# Enable or disable extensions
# See https://docs.scrapy.org/en/latest/topics/extensions.html
# EXTENSIONS = {
#    'scrapy.extensions.telnet.TelnetConsole': None,
# }

# Configure item pipelines
# See https://docs.scrapy.org/en/latest/topics/item-pipeline.html
# ITEM_PIPELINES = {
#    'matchreports.pipelines.MatchreportsPipeline': 300,
# }

# Enable and configure the AutoThrottle extension (disabled by default)
# See https://docs.scrapy.org/en/latest/topics/autothrottle.html
# AUTOTHROTTLE_ENABLED = True
# The initial download delay
# AUTOTHROTTLE_START_DELAY = 5
# The maximum download delay to be set in case of high latencies
# AUTOTHROTTLE_MAX_DELAY = 60
# The average number of requests Scrapy should be sending in parallel to
# each remote server
# AUTOTHROTTLE_TARGET_CONCURRENCY = 1.0
# Enable showing throttling stats for every response received:
# AUTOTHROTTLE_DEBUG = False

# Enable and configure HTTP caching (disabled by default)
# See https://docs.scrapy.org/en/latest/topics/downloader-middleware.html#httpcache-middleware-settings
# HTTPCACHE_ENABLED = True
# HTTPCACHE_EXPIRATION_SECS = 0
# HTTPCACHE_DIR = 'httpcache'
# HTTPCACHE_IGNORE_HTTP_CODES = []
# HTTPCACHE_STORAGE = 'scrapy.extensions.httpcache.FilesystemCacheStorage'


matchreports/matchreports/middlewares.py
import logging
import os
import time
import re
from datetime import datetime

from scrapy.http import Request
from scrapy.item import Item
from scrapy.utils.request import request_fingerprint
from scrapy.utils.project import data_path
from scrapy.utils.python import to_bytes
from scrapy.exceptions import NotConfigured
from scrapy import signals
import sqlite3

logger = logging.getLogger(__name__)

# Define here the models for your spider middleware
#
# See documentation in:
# https://docs.scrapy.org/en/latest/topics/spider-middleware.html

from scrapy import signals

# useful for handling different item types with a single interface
from itemadapter import is_item, ItemAdapter


class MatchreportsSpiderMiddleware:
    # Not all methods need to be defined. If a method is not defined,
    # scrapy acts as if the spider middleware does not modify the
    # passed objects.

    @classmethod
    def from_crawler(cls, crawler):
        # This method is used by Scrapy to create your spiders.
        s = cls()
        crawler.signals.connect(s.spider_opened, signal=signals.spider_opened)
        return s

    def process_spider_input(self, response, spider):
        # Called for each response that goes through the spider
        # middleware and into the spider.

        # Should return None or raise an exception.
        return None

    def process_spider_output(self, response, result, spider):
        # Called with the results returned from the Spider, after
        # it has processed the response.

        # Must return an iterable of Request, or item objects.
        for i in result:
            yield i

    def process_spider_exception(self, response, exception, spider):
        # Called when a spider or process_spider_input() method
        # (from other spider middleware) raises an exception.

        # Should return either None or an iterable of Request or item objects.
        pass

    def process_start_requests(self, start_requests, spider):
        # Called with the start requests of the spider, and works
        # similarly to the process_spider_output() method, except
        # that it doesn’t have a response associated.

        # Must return only requests (not items).
        for r in start_requests:
            yield r

    def spider_opened(self, spider):
        spider.logger.info("Spider opened: %s" % spider.name)


class MatchreportsDownloaderMiddleware:
    # Not all methods need to be defined. If a method is not defined,
    # scrapy acts as if the downloader middleware does not modify the
    # passed objects.

    @classmethod
    def from_crawler(cls, crawler):
        # This method is used by Scrapy to create your spiders.
        s = cls()
        crawler.signals.connect(s.spider_opened, signal=signals.spider_opened)
        return s

    def process_request(self, request, spider):
        # Called for each request that goes through the downloader
        # middleware.

        # Must either:
        # - return None: continue processing this request
        # - or return a Response object
        # - or return a Request object
        # - or raise IgnoreRequest: process_exception() methods of
        #   installed downloader middleware will be called
        return None

    def process_response(self, request, response, spider):
        # Called with the response returned from the downloader.

        # Must either;
        # - return a Response object
        # - return a Request object
        # - or raise IgnoreRequest
        return response

    def process_exception(self, request, exception, spider):
        # Called when a download handler or a process_request()
        # (from other downloader middleware) raises an exception.

        # Must either:
        # - return None: continue processing this exception
        # - return a Response object: stops process_exception() chain
        # - return a Request object: stops process_exception() chain
        pass

    def spider_opened(self, spider):
        spider.logger.info("Spider opened: %s" % spider.name)


class DeltaFetch(object):
    """
    This is a spider middleware to ignore requests to pages containing items
    seen in previous crawls of the same spider, thus producing a "delta crawl"
    containing only new items.
    This also speeds up the crawl, by reducing the number of requests that need
    to be crawled, and processed (typically, item requests are the most cpu
    intensive).
    """

    def __init__(self, dir, reset=False, stats=None):
        self.db = None
        self.dir = dir
        self.reset = reset
        self.stats = stats

    @classmethod
    def from_crawler(cls, crawler):
        s = crawler.settings
        if not s.getbool("DELTAFETCH_ENABLED"):
            raise NotConfigured
        dir = data_path(s.get("DELTAFETCH_DIR", "deltafetch"))
        reset = s.getbool("DELTAFETCH_RESET")
        o = cls(dir, reset, crawler.stats)
        crawler.signals.connect(o.spider_opened, signal=signals.spider_opened)
        crawler.signals.connect(o.spider_closed, signal=signals.spider_closed)
        return o

    def spider_opened(self, spider):
        if not os.path.exists(self.dir):
            os.makedirs(self.dir)
        dbpath = os.path.join(self.dir, f"{spider.name}.db")
        reset = self.reset or getattr(spider, "deltafetch_reset", False)
        try:
            self.db = kvstore(dbpath, reset=reset)
        except Exception:
            logger.warning(f"Failed to open DeltaFetch database at {dbpath}, trying to recreate it")
            if os.path.exists(dbpath):
                os.remove(dbpath)
            self.db = kvstore(dbpath, reset=reset)

    def spider_closed(self, spider):
        self.db.close()

    def process_spider_output(self, response, result, spider):
        for r in result:
            if isinstance(r, Request):
                key = self._get_key(r)
                val = self._get_val(r)
                if key in self.db.keys():
                    logger.info(f"Ignoring already visited: {r.url}")
                    if self.stats:
                        self.stats.inc_value("deltafetch/skipped", spider=spider)
                    continue
            elif isinstance(r, (Item, dict)):
                key = self._get_key(response.request)
                val = self._get_val(response.request)
                self.db[key] = val
                if self.stats:
                    self.stats.inc_value("deltafetch/stored", spider=spider)
            yield r

    def _get_key(self, request):
        key = request.meta.get("deltafetch_key") or request_fingerprint(request)
        # request_fingerprint() returns `hashlib.sha1().hexdigest()`, is a string
        return to_bytes(key)

    def _get_val(self, request):
        # if re.search("\d{4}\/[a-z]{3}\/\d{2}", request.url):
        #    date = re.search("\d{4}\/[a-z]{3}\/\d{2}", request.url).group()
        #    date = datetime.strptime(date, "%Y/%b/%d").strftime("%Y-%m-%d")
        # else:
        #    date = time.time()
        return time.time()


class kvstore(dict):
    def __init__(self, filename=None, reset=False):
        self.conn = sqlite3.connect(filename)
        if reset:
            del_table()
        self.conn.execute("CREATE TABLE IF NOT EXISTS kv (key text unique, value timestamp)")

    def close(self):
        self.conn.commit()
        self.conn.close()

    def del_table(self):
        self.conn.execute("DROP TABLE [IF EXISTS] kv")

    def __len__(self):
        rows = self.conn.execute("SELECT COUNT(*) FROM kv").fetchone()[0]
        return rows if rows is not None else 0

    def iterkeys(self):
        c = self.conn.cursor()
        for row in self.conn.execute("SELECT key FROM kv"):
            yield row[0]

    def itervalues(self):
        c = self.conn.cursor()
        for row in c.execute("SELECT value FROM kv"):
            yield row[0]

    def iteritems(self):
        c = self.conn.cursor()
        for row in c.execute("SELECT key, value FROM kv"):
            yield row[0], row[1]

    def keys(self):
        return list(self.iterkeys())

    def values(self):
        return list(self.itervalues())

    def items(self):
        return list(self.iteritems())

    def less_days_max_value(self, val, days: int = 14):
        return self.conn.execute(f"SELECT date({val}, '-{days} day') < MAX(value) FROM kv").fetchone()[0]

    def __contains__(self, key):
        return self.conn.execute("SELECT 1 FROM kv WHERE key = ?", (key,)).fetchone() is not None

    def __getitem__(self, key):
        item = self.conn.execute("SELECT value FROM kv WHERE key = ?", (key,)).fetchone()
        if item is None:
            raise KeyError(key)
        return item[0]

    def __setitem__(self, key, value):
        self.conn.execute("REPLACE INTO kv (key, value) VALUES (?,?)", (key, value))
        self.conn.commit()

    def __delitem__(self, key):
        if key not in self:
            raise KeyError(key)
        self.conn.execute("DELETE FROM kv WHERE key = ?", (key,))
        self.conn.commit()

    def __iter__(self):
        return self.iterkeys()

matchreports/matchreports/spiders/guardian.py
import logging
import os
import re
from datetime import datetime
from pathlib import Path

from twisted.internet import reactor
import scrapy
from scrapy.loader import ItemLoader

from matchreports.settings import guardiansitesettings, settings
from matchreports.items import MatchreportsItem


def extract_attrib(img_objs):
    return [{k: v for k, v in img_obj.attrib.items()} for img_obj in img_objs]


class GuardianSpider(scrapy.Spider):
    name = "guardian-match-reports"

    def __init__(self, *args, **kwargs):
        self.timeout = int(kwargs.pop("timeout", "60"))
        super(GuardianSpider, self).__init__(*args, **kwargs)

    def start_requests(self):
        reactor.callLater(self.timeout, self.stop)

        urls = [guardiansitesettings.BASE_URL]
        for url in urls:
            yield scrapy.Request(url=url, callback=self.parse)

    def parse(self, response):
        matchreport_page_links = response.css(".js-headline-text::attr(href)").getall()
        yield from response.follow_all(matchreport_page_links, self.parse_matchreport)

        next_page = response.css(".pagination__text+ .pagination__action--static::attr(href)").get()
        if next_page is not None:
            yield response.follow(next_page, self.parse)

    def parse_matchreport(self, response):
        date = re.search("\d{4}\/[a-z]{3}\/\d{2}", response.url).group()
        date = datetime.strptime(date, "%Y/%b/%d")
        teams = response.css(".css-184iqxr .css-w5gu9i::text").getall()
        self.log(f"Following link to page {response.url}")
        pictures = extract_attrib(response.css("img"))
        guardian_loader = ItemLoader(MatchreportsItem(), response)
        guardian_loader.add_value("scrape_date", datetime.now())
        guardian_loader.add_value("link", response.url)
        guardian_loader.add_value("headline", response.css(".css-dxy9hs::text").get())
        guardian_loader.add_value("home_team", teams[0])
        guardian_loader.add_value("away_team", teams[1])
        guardian_loader.add_value("match_date", date)
        guardian_loader.add_value("author", response.css(".css-lkdvty a::text").get())
        guardian_loader.add_value("stadium", response.css(".css-lkdvty::text").re("(?=[A-Z]).*")[0])
        guardian_loader.add_value("text", response.css(".css-19th2c::text").getall())
        guardian_loader.add_value("pictures", pictures)
        yield guardian_loader.load_item()

    def stop(self):
        self.crawler.engine.close_spider(self, "timeout")


if __name__ == "__main__":
    from scrapy.crawler import CrawlerProcess
    from scrapy.utils.project import get_project_settings

    settings = get_project_settings()
    settings["FEEDS"] = {f"s3://{settings.BUCKET_NAME}/scraping/feeds/%(name)s.jl": {"format": "jl"}}
    process = CrawlerProcess(settings=settings)
    process.crawl(GuardianSpider)
    process.start()

src/spider.py
import logging
import os
import re
from datetime import datetime
from pathlib import Path

from twisted.internet import reactor
import scrapy

from settings import guardiansitesettings, settings

# TODO check error
# TODO use items, item loaders and default values to fix  errors.
# use  item_loader.add_value to  use  existing methods and probabaly create try except stuff
# or alternatively add the img attrib  stuff with add_css  and some filters from scrapy
# TODO  set some values to decrease banning prob. see bottom
# TODO out to s3
# TODO never repopulate (using a latest matchdate variable and a populated links in the data storage path)
# https://stackoverflow.com/questions/29396942/avoiding-scraping-data-from-pages-already-scraped
# TODO possibly make a statscollector to  check skipped pages or  so and use it to stop crawling


class GuardianSpider(scrapy.Spider):
    name = "guardian-match-reports"

    def __init__(self, *args, **kwargs):
        self.timeout = int(kwargs.pop("timeout", "60"))
        super(GuardianSpider, self).__init__(*args, **kwargs)

    def start_requests(self):
        reactor.callLater(self.timeout, self.stop)

        urls = [guardiansitesettings.BASE_URL]
        for url in urls:
            yield scrapy.Request(url=url, callback=self.parse)

    def parse(self, response):
        matchreport_page_links = response.css(".js-headline-text::attr(href)").getall()
        yield from response.follow_all(matchreport_page_links, self.parse_matchreport)

        next_page = response.css(".pagination__action--static::attr(href)").get()
        if next_page is not None:
            yield response.follow(next_page, self.parse)

    def parse_matchreport(self, response):
        date = re.search("\d{4}\/[a-z]{3}\/\d{2}", response.url).group()
        date = datetime.strptime(date, "%Y/%b/%d").strftime("%Y-%m-%d")
        teams = response.css(".css-184iqxr .css-w5gu9i::text").getall()
        self.log(f"Following link to page {response.url}")
        yield {
            "link": response.url,
            "headline": response.css(".css-dxy9hs::text").get(),
            "home_team": teams[0],
            "away_team": teams[1],
            "match_date": date,
            "pictures": [
                {"link": img.attrib["src"], "text": img.attrib["alt"]} for img in response.css("img") if "alt" in img.attrib.keys()
            ],
            "author": response.css(".css-lkdvty a::text").get(),
            "stadium": response.css(".css-lkdvty::text").re("(?=[A-Z]).*")[0],
            "text": response.css(".css-19t0h2c::text").getall(),
        }

    def stop(self):
        self.crawler.engine.close_spider(self, "timeout")


if __name__ == "__main__":
    from scrapy.crawler import CrawlerProcess

    process = CrawlerProcess(
        settings={
            "FEEDS": {
                f"s3://{settings.BUCKET_NAME}/scraping/feeds/%(name)s.jl": {"format": "json"},
            },
        }
    )

    process.crawl(GuardianSpider)
    process.start()

#   File "/Users/hojland/guardian_soccer_spider/src/spider.py", line 50, in <listcomp>
#    {"link": img.attrib["src"], "text": img.attrib["alt"]} for img in response.css("img") if "alt" in img.attrib.keys()
# KeyError: 'src'

# 2021-04-05 22:05:54 [scrapy.core.scraper] ERROR: Spider error processing <GET https://www.theguardian.com/football/2021/mar/07/manchester-city-manchester-united-premier-league-match-report> (referer: https://www.theguardian.com/football/premierleague+tone/matchreports?page=2)
#  2021-04-05 22:05:54 [scrapy.core.scraper] ERROR: Spider error processing <GET https://www.theguardian.com/football/2021/mar/06/burnley-arsenal-premier-league-report> (referer: https://www.theguardian.com/football/premierleague+tone/matchreports?page=2)

# rotate your user agent from a pool of well-known ones from browsers (google around to get a list of them)
# disable cookies (see COOKIES_ENABLED) as some sites may use cookies to spot bot behaviour
# use download delays (2 or higher). See DOWNLOAD_DELAY setting.
# if possible, use Google cache to fetch pages, instead of hitting the sites directly
# use a pool of rotating IPs. For example, the free Tor project or paid services like ProxyMesh. An open source alternative is scrapoxy, a super proxy that you can attach your own proxies to.
# use a highly distributed downloader that circumvents bans internally, so you can just focus on parsing clean pages. One example of such downloaders is Crawlera